jenkins/runlocal

```diff
@@ -21,6 +21,7 @@ make -C server/rpm clean rpm
 # $ buildah login -u='$app' -p="${__LOGIN_SECRET__}" ${PB_CONTAINER_REG}
 
 RPM_PATH=${HOME}/rpmbuild/RPMS/noarch/pbench-server-*.rpm bash -ex ./server/pbenchinacan/container-build-ci.sh
+buildah push localhost/pbench-server:${PB_SERVER_IMAGE_TAG} ${PB_CONTAINER_REG}/${PB_ORG_NAME}/pbench-server:${PB_SERVER_IMAGE_TAG}
```

Comment:
This is already in `main`, right?  (If not, it was omitted from another PR and slipped in here.... 🙃)

Reply:
We need both the stand-alone and the in-a-can image pushed, but `runlocal` tags them with `${USER}` instead of the branch, and I just updated this to not use the same image name as the CI to avoid confusion.

Reply:
> Why do we need this script (which, ostensibly, is for _local_ execution) to push the images to a **remote** registry?

Because I don't know of a way to use "localhost" as a local registry the same way we use the configured registry host.

> If we don't have to push the images, then we don't have to worry too much about how they are named and tagged.

This would be ideal.  Just have the CI create a local image and run it in a pod without the use of a registry, and then just push the final image once all the tests pass.

> And, we don't have to worry about managing a proliferation of stuff on the remote registry.

Agreed.

> I'm starting to feel like we need to reexamine what we mean by the "image name". When we push an image to `${PB_CONTAINER_REG}/${PB_ORG_NAME}/${PB_SERVER_IMAGE_NAME}:${PB_SERVER_IMAGE_TAG}`, the `${PB_ORG_NAME}/${PB_SERVER_IMAGE_NAME}` portion is, I believe the _repository_ (with the part before that being the _registry_), and there is only ever one image per repository, with different versions of it referenced by different tags.

Okay.

> So, we shouldn't need to have both `pbench-server-ci` and `pbench-server-local` repositories...if we select the tag values properly. (How many repositories do we want to have?)

Agreed, those images should be just different labels.

> Perhaps instead of adding `-local` we should be dropping the `-ci`? As things stand currently, there's no collision between the CI and `runlocal`, because the CI uses PR numbers and branch names for tags, while `runlocal` uses `${USER}`.

I don't think `pbench-server` is the same as `pbench-server-ci`.  One is built by `container-build-sa.sh` and one by `container-build-ci.sh`, with the second built on top of the first.

Reply:
> I don't know of a way to use "localhost" as a local registry the same way we use the configured registry host.

Perhaps if we moved away from our heavily-component-ized image name and moved back toward something more single-token, we could just let the scripts define what they want (e.g., if we can settle on what the repository should be called, then we can let the prefix be either `quay.io/pbench` or `localhost`, and then they are swappable).

Alternatively, I believe that you can just use `localhost` where you would otherwise use something like `quay.io`:
```
$ podman tag b25b72337878 localhost/pbench/pbench-agent-all-fedora-36 
$ podman image ls
REPOSITORY                                               TAG               IMAGE ID      CREATED        SIZE
localhost/pbench/pbench-agent-all-fedora-36              latest            b25b72337878  4 weeks ago    2.06 GB
[...]
```
Note the `pbench` "organization" in the `localhost` "registry".

And we have more options than that:  you can have an image in your local registry named (e.g.) `quay.io/pbench/pbench-ci-fedora` which you built locally.  You can run that image locally.  And, unless you `push` it (or pull it), it will never be associated with anything actually in the remote `quay.io` registry.  (The Podman `tag` command can be used to set an image name, independent of the actual repository that initially came from; this allows you to set a default destination for the `push`; and, I think you can set the full image name via the `-t` flag at build-time.)

So, in `container-build-ci.sh` we _could_ build the image and tag it as `${PB_CONTAINER_REG}/${PB_ORG_NAME}/${PB_SERVER_IMAGE_NAME}:${PB_SERVER_IMAGE_TAG}` _before_ we push it (instead of using `localhost`); then it would be available locally _with that name_ without having to be pushed, first.  That would allow the CI to push it and `runlocal` to not push it.  (And, if we manage the tags properly, then we don't need to worry about the pull policy messing things up for local builds.)

> Just have the CI create a local image and run it in a pod without the use of a registry, and then just push the final image once all the tests pass.

I had myself convinced that this was a bad idea, which is why the CI always pushes.  I don't remember the details now, but I decided that pushing the image between building and running avoided some problems (I didn't want to require the tests to run on the same executor as the one which did the container build, in case we wanted to run parallel stages).  Also, in the case of PRs, I think the image should be pushed if the tests fail, so that the developer can easily investigate things.  And, the same is likely true for the branch builds.

> I don't think `pbench-server` is the same as `pbench-server-ci`.

I think they are "close enough", but if you really think they should be in separate repos, I'm OK with that.  But, I think we should be able to settle on the names well enough to hard-code them.  Perhaps `pbench-server` is OK as is (my attempts to improve it...didn't); perhaps we could replace `pbench-server-ci` with `pbench-server-in-a-can`?

Reply:
> > I don't know of a way to use "localhost" as a local registry the same way we use the configured registry host.
> 
> Perhaps if we moved away from our heavily-component-ized image name and moved back toward something more single-token, we could just let the scripts define what they want (e.g., if we can settle on what the repository should be called, then we can let the prefix be either `quay.io/pbench` or `localhost`, and then they are swappable).
> 
> Alternatively, I believe that you can just use `localhost` where you would otherwise use something like `quay.io`:
> 
> ```
> $ podman tag b25b72337878 localhost/pbench/pbench-agent-all-fedora-36 
> $ podman image ls
> REPOSITORY                                               TAG               IMAGE ID      CREATED        SIZE
> localhost/pbench/pbench-agent-all-fedora-36              latest            b25b72337878  4 weeks ago    2.06 GB
> [...]
> ```
> 
> Note the `pbench` "organization" in the `localhost` "registry".
> 
> And we have more options than that: you can have an image in your local registry named (e.g.) `quay.io/pbench/pbench-ci-fedora` which you built locally. You can run that image locally. And, unless you `push` it (or pull it), it will never be associated with anything actually in the remote `quay.io` registry. (The Podman `tag` command can be used to set an image name, independent of the actual repository that initially came from; this allows you to set a default destination for the `push`; and, I think you can set the full image name via the `-t` flag at build-time.)
> 
> So, in `container-build-ci.sh` we _could_ build the image and tag it as `${PB_CONTAINER_REG}/${PB_ORG_NAME}/${PB_SERVER_IMAGE_NAME}:${PB_SERVER_IMAGE_TAG}` _before_ we push it (instead of using `localhost`); then it would be available locally _with that name_ without having to be pushed, first. That would allow the CI to push it and `runlocal` to not push it. (And, if we manage the tags properly, then we don't need to worry about the pull policy messing things up for local builds.)

I love how all this sounds, but I have run afoul of not being able to use `localhost` as a registry in the past.  If we can make this work for us, great.  I am still doubtful it'll work the way we want.

> > Just have the CI create a local image and run it in a pod without the use of a registry, and then just push the final image once all the tests pass.
> 
> I had myself convinced that this was a bad idea, which is why the CI always pushes. I don't remember the details now, but I decided that pushing the image between building and running avoided some problems (I didn't want to require the tests to run on the same executor as the one which did the container build, in case we wanted to run parallel stages). Also, in the case of PRs, I think the image should be pushed if the tests fail, so that the developer can easily investigate things. And, the same is likely true for the branch builds.

I agree with the failure mode pushes, that is a good point.

> Perhaps `pbench-server` is OK as is (my attempts to improve it...didn't); perhaps we could replace `pbench-server-ci` with `pbench-server-in-a-can`?

That sounds reasonable to me.

Reply:
So then let's merge this and work out the improvements in a separate PR.